rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,367 @@
+upload status. This additional functionality can be built into the existing DP Backend Lambda.
+
+The backend will also be responsible for performing a simple validation of the file being uploaded. The validation is
+performed during the initial request to upload. The validation consists of verifying the extension of the file is a [supported type](https://github.com/chanzuckerberg/corpora-data-portal/blob/main/backend/schema/corpora_schema.md#implementations).
+
+#### Upload Queue
+
+The Upload Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending upload jobs.
+The upload jobs are used by the upload service to know where to download the file from before uploading them to S3.
+For the first implementation, the job sleep time will be 15min. Eventually, the job sleep time will depend on the size
+of the file being uploaded.
+
+Cloudwatch events can be used to monitor the SQS length. Additional compute resource can be spawned in response to an
+increase in jobs in the queue.
+
+##### Upload Job Entry
+
+These are the fields that will be in the upload job placed in the queue:
+
+- Link - the link to file to download. Use to retrieve the file from DropBox.
+- submission_id - identifies the submission. Used to determine the storage location.
+- dataset_id - identifies the dataset. Used to determine the storage location.
+- file_name - the name of the file being downloaded.
+
+#### Upload Service
+
+A computing resource within the AWS cloud. It will receive upload jobs from the Upload Queue. Those jobs contain
+the information needed by the compute resource to download a file from DropBox and upload it to S3. The computing resource
+will also verify the integrity of the download and upload. The compute resource will also change the state of the
+upload in the Upload table as needed.
+
+While the upload job is processing, the compute resource with occasionally poll the upload table for a cancel
+pending status. If the cancel pending status is set, the compute resource will delete any data uploaded to S3,
+remove the upload job from the queue, and change the status of the upload to canceled.
+
+If an unrecoverable error occurs while processing the upload, the compute resource will return the upload job to the
+queue and update the upload table.
+
+If an error is encountered that cannot be fixed by retrying, the upload status will change to failed. The error can be
+returned
+
+##### Upload Table
+
+The upload table is a DynamoDB table that tracks the current status of uploads. The table will have the following columns:
+
+- dataset_id - primary key
+- status - the current state of the upload.
+- progress - the current progress of the upload.
+- owner - the id of the user who owns this collection.
+- message - error message if the upload fails.
+
+#### Upload Bucket
+
+The upload bucket is a separate bucket or separate bucket location within our main bucket. It is used to stage files
```

Comment:
What's the difference between "separate bucket" and "separate bucket location"?

Reply:
Also, after the upload is complete, there will be another set of processes the kick off (i.e. validation, copy to final S3 destination, etc.), how will they know which bucket to look into? Are you updating the RDS Project table with the S3 bucket location (I believe there's a field there already).

Reply:
We will use a separate bucket for uploading the files. I talked with @mckinsel and he said he planned on using s3 eventing to kick off his validation workflows. He will handle putting the files in the main bucket and updating the DB with the file locations. I added a section about handing off to validation.